### PR TITLE
HDDS-15600. Fix ListObjects response for encoding-type and empty delimiter

### PR DESCRIPTION
**Pull request by Gargi-jais11**
## What changes were proposed in this pull request?
Ozone S3 Gateway fails s3-tests ListObjects cases that AWS S3 passes:

- `test_bucket_list_encoding_basic` — With EncodingType=url, keys and CommonPrefixes containing spaces are encoded with + (Java URLEncoder form encoding) instead of AWS-style %20.
Example: prefix **quux ab/** is returned as **quux+ab/** but should be **quux%20ab/**.
- `test_bucket_list_delimiter_empty` — When **Delimiter=''** is sent, listing behavior is correct (all keys returned, no CommonPrefixes), but the response incorrectly includes a Delimiter field. AWS omits Delimiter from the XML when the client passes an empty delimiter.

> - `test_bucket_list_prefix_unreadable` — ListObjects with **Prefix='\x0a' (newline)** should echo the prefix in the response and return empty Contents/CommonPrefixes. Ozone may not preserve or echo the control-character prefix correctly.
No conclusion is made on the root cause of this test failure. Skipping it to be fixed in this PR.

https://ozone.s3.peterxcli.dev/#latest-run-section

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-15600

## How was this patch tested?

**Before fix:**
1. encoding type:
```
bash-5.1$ aws s3api list-objects   --bucket buck1   --encoding-type url   --delimiter /   --endpoint-url http://s3g:9878/
{
    "Contents": [
        {
            "Key": "asdf%2Bb",
            "LastModified": "2026-06-19T09:51:39.569Z",
            "ETag": "\"f75b8179e4bbe7e2b4a074dcef62de95\"",
            "Size": 8,
            "StorageClass": "STANDARD",
            "Owner": {
                "DisplayName": "testuser",
                "ID": "bb2bd7ca4a327f84e6cd3979f8fa3828a50a08893c1b68f9d6715352c8d07b93"
            }
        }
    ],
    "CommonPrefixes": [
        {
            "Prefix": "foo/"
        },
        {
            "Prefix": "foo%2B1/"
        },
        {
            "Prefix": "quux+ab/".                       <----------------- wrong output
        }
    ],
    "RequestCharged": null,
    "Prefix": ""
}
```
2. empty delimiter:
```
bash-5.1$ aws --debug s3api list-objects \
  --bucket buck1 \
  --delimiter "" \
  --endpoint-url http://s3g:9878 2>&1 \
  | grep -oE '<Prefix>[^<]*</Prefix>|<Delimiter>[^<]*</Delimiter>|<KeyCount>[^<]*</KeyCount>'
<Prefix></Prefix>
<KeyCount>0</KeyCount>
<Delimiter></Delimiter>.           <---------- should not be present
```

**After fix:**
1. encoding-type:
```
bash-5.1$ aws s3api list-objects   --bucket buck1   --encoding-type url   --delimiter /   --endpoint-url http://s3g:9878/
{
    "Contents": [
        {
            "Key": "asdf%2Bb",
            "LastModified": "2026-06-19T09:08:58.933Z",
            "ETag": "\"f75b8179e4bbe7e2b4a074dcef62de95\"",
            "Size": 8,
            "StorageClass": "STANDARD",
            "Owner": {
                "DisplayName": "testuser",
                "ID": "bb2bd7ca4a327f84e6cd3979f8fa3828a50a08893c1b68f9d6715352c8d07b93"
            }
        }
    ],
    "CommonPrefixes": [
        {
            "Prefix": "foo/"
        },
        {
            "Prefix": "foo%2B1/"
        },
        {
            "Prefix": "quux%20ab/"                <------------------------correct output
        }
    ],
    "RequestCharged": null,
    "Prefix": null
}
```
2. empty delimeter :
```
bash-5.1$ aws --debug s3api list-objects \
  --bucket buck1 \
  --delimiter "" \
  --endpoint-url http://s3g:9878 2>&1 \
  | grep -oE '<Prefix>[^<]*</Prefix>|<Delimiter>[^<]*</Delimiter>|<KeyCount>[^<]*</KeyCount>'
  
<KeyCount>0</KeyCount>.                 <-------------------- no empty delimiter present
```

### CODE REVIEW

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/util/S3Utils.java [hunk 1]
```diff
@@ -62,6 +62,17 @@ public static String urlEncode(String str)
...
+  /**
+   * Percent-encode a string for S3 {@code encoding-type=url} responses.
+   *
+   * <p>Unlike {@link URLEncoder} (application/x-www-form-urlencoded), AWS S3
+   * uses percent-encoding where spaces are {@code %20}, not {@code +}.
+   */
+  public static String s3urlEncode(String str)
+      throws UnsupportedEncodingException {
+    return urlEncode(str).replace("+", "%20");
+  }
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/util/S3Utils.java:74]: Can we use `PercentCodec` from `commons-codec`?

https://commons.apache.org/proper/commons-codec/apidocs/org/apache/commons/codec/net/PercentCodec.html

**Review by adoroszlai — COMMENTED** [review]
Thanks @Gargi-jais11 for the patch.

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/util/S3Utils.java (hunk 1 above)
Comment (Gargi-jais11) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/util/S3Utils.java:74]: Okay let me check and get back to u.

Comment (Gargi-jais11) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/util/S3Utils.java:74]: @adoroszlai I need to determine whether `PercentCodec` produces the same encoding semantics that AWS S3 expects. Currently I am gong through some of the documentation. If there is no issue then I will update it.

Comment (Gargi-jais11) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/util/S3Utils.java:74]: @adoroszlai 
On the basis of local testing I found that`PercentCodec` alone does not encode ASCII specials like **=,** which S3 requires. Using it correctly needs configuring **alwaysEncodeChars** to mirror URLEncoder's unsafe set. Since the only S3 difference from URLEncoder is %20 vs +, urlEncode(str).replace("+", "%20") is simpler and equivalent.

Comment (adoroszlai) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/util/S3Utils.java:74]: Thanks for checking.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java [hunk 2]
```diff
@@ -1095,11 +1095,11 @@ public ListKeysResult listKeys(String volumeName, String bucketName,
     } else {
       // This allows us to seek directly to the first key with the right prefix.
       seekKey = getOzoneKey(volumeName, bucketName,
-          StringUtils.isNotBlank(keyPrefix) ? keyPrefix : OM_KEY_PREFIX);
+          (keyPrefix != null && !keyPrefix.isEmpty()) ? keyPrefix : OM_KEY_PREFIX);
```
Comment (chungen0126) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java]: We can use `StringUtils.isNotEmpty()` to simplify this condition.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java [hunk 3]
```diff
@@ -1095,11 +1095,11 @@ public ListKeysResult listKeys(String volumeName, String bucketName,
...
     }
 
     String seekPrefix;
-    if (StringUtils.isNotBlank(keyPrefix)) {
+    if (keyPrefix != null && !keyPrefix.isEmpty()) {
```
Comment (chungen0126) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java]: We can use `StringUtils.isNotEmpty()` to simplify this condition.

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java [hunk 4]
```diff
@@ -157,17 +159,21 @@ Response handleGetRequest(S3RequestContext context, String bucketName) throws IO
...
-    //   Delimiter, Prefix, Key, and StartAfter.
+    // encoded key name values in the following response elements: Delimiter,
+    // Key, and StartAfter. The echoed Prefix request parameter is returned without URL encoding.
     //
     // For detail refer:
     // https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#AmazonS3-ListObjectsV2-response-EncodingType
     ListObjectResponse response = new ListObjectResponse();
-    response.setDelimiter(EncodingTypeObject.createNullable(delimiter, encodingType));
+    // AWS omits Delimiter from the response when the client passes delimiter= or does not specify delimiter at all.
+    if (delimiter != null && !delimiter.isEmpty()) {
```
Comment (chungen0126) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java]: We can use StringUtils.isNotEmpty() to simplify this condition.

**Review by chungen0126 — APPROVED** [review]
Thanks @Gargi-jais11 for working on this. Overall, looks good to me. Just a nit.

**Comment by Gargi-jais11** [comment]
@ivandika3 please review this patch.

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java [hunk 5]
```diff
@@ -157,17 +159,21 @@ Response handleGetRequest(S3RequestContext context, String bucketName) throws IO
...
-    response.setDelimiter(EncodingTypeObject.createNullable(delimiter, encodingType));
+    // AWS omits Delimiter from the response when the client passes delimiter= or does not specify delimiter at all.
+    if (StringUtils.isNotEmpty(delimiter)) {
+      response.setDelimiter(EncodingTypeObject.createNullable(delimiter, encodingType));
+    }
     response.setName(bucketName);
-    response.setPrefix(EncodingTypeObject.createNullable(prefix, encodingType));
+    if (prefixSpecified) {
+      response.setPrefix(EncodingTypeObject.createNullable(prefix, null));
+    }
```
Comment (Russole) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java]: `Prefix` should also be encoded when `encoding-type=url` is specified.

AWS docs say `EncodingType` applies to `Delimiter`, `Prefix`, `Key`, and `StartAfter`:
https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#AmazonS3-ListObjectsV2-response-EncodingType

Local test:

```bash
aws s3api list-objects \
  --bucket buck1 \
  --prefix $'\n' \
  --encoding-type url \
  --endpoint-url http://localhost:9878/
```

Actual:

```json
"Prefix": "\n"
```

Expected:

```json
"Prefix": "%0A"
```

Looks like `Prefix` is created with a `null` encoding type here. Should this use `encodingType` instead?

**Review by Russole — CHANGES_REQUESTED** [review]
Thanks @Gargi-jais11 for the patch. I’ve left a few review comments.

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java (hunk 5 above)
Comment (Gargi-jais11) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java]: @Russole thanks for bringing this in view. Will revert the changes for this but now the problem is `test_bucket_list_prefix_unreadable` this test since the encoding type is not there so it should not encode it. 

With original :
```
if (prefixSpecified) {
      response.setPrefix(EncodingTypeObject.createNullable(prefix, encodingType));
    }
```
**Local test:**
```
bash-5.1$ aws s3api list-objects \
  --bucket buck1 \
  --prefix $'\n' \
  --endpoint-url http://s3g:9878/
```
**Actual is:**
```
{
    "RequestCharged": null,
    "Prefix": "%0A"            --------------------> it is encoded without encoding type mentioned in the request
}
```

**Expected :**
```
{
    "RequestCharged": null,
    "Prefix": "\n"
}
```

Comment (Gargi-jais11) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java]: I wonder how to fix this part when encoding-type is not mentioned in the request.

Comment (Russole) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java]: If `queryParams().get(QueryParams.PREFIX)` already returns the decoded value `"\n"`, then no extra request-decoding fix is needed here.

Passing `encodingType` should handle both cases:

- when `encodingType == null`, `Prefix` is marshalled as `"\n"`
- when `encodingType == "url"`, `Prefix` is marshalled as `"%0A"`

So using `EncodingTypeObject.createNullable(prefix, encodingType)` should not encode `Prefix` when `encoding-type` is absent.

Comment (Gargi-jais11) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java]: but both the cases are not handled properly in case of null encoding-type its giving **%0A**.

Comment (Gargi-jais11) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java]: To identify the root cause of this, I have added few logs around`response.setPrefix(EncodingTypeObject.createNullable(prefix, encodingType));`

**with encoding-type**: correct response
```
bash-5.1$ aws s3api list-objects \
  --bucket buck1 \
  --prefix $'\n' \
  --encoding-type url \
  --endpoint-url http://s3g:9878/
{
    "RequestCharged": null,
    "Prefix": "%0A"
}
```
Related logs for this request:
```
prefixSpecified=true, prefix='\n', encodingType='url'

createNullable: value='\n', encodingType='url'

EncodingTypeObject ctor: value='\n', encodingType='url'

prefixObj.value='\n'
prefixObj.encodingType='url'
"GET /buck1?encoding-type=url&prefix=%0A HTTP/1.1" 200 304
```

**Without Encoding-Type**: output is wrong
```
bash-5.1$ aws s3api list-objects   --bucket buck1   --prefix $'\n'   --endpoint-url http://s3g:9878/
{
    "RequestCharged": null,
    "Prefix": "%0A"
}
```

Related logs to this commands:
```
prefixSpecified=true, prefix='\n', encodingType='url'

createNullable: value='\n', encodingType='url'

EncodingTypeObject ctor: value='\n', encodingType='url'

prefixObj.value='\n'
prefixObj.encodingType='url'

// the HTTP request logged by S3G is:
"GET /buck1?prefix=%0A&encoding-type=url HTTP/1.1" 200 304
```

This is the important finding:

The second request was sent without `--encoding-type url`, but S3G is processing a request that contains encoding-type=url.

Because of that:
- encodingType is never null.
- EncodingTypeObject.createNullable() is always invoked with "url".
- The response is therefore always URL-encoded.

Comment (Gargi-jais11) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java]: @Russole for now I have reverted the fix as this URL-encoded seems to be coming from aws s3 itself. Which is verified from the above experiment which I did locally. We can come up for this test fix later. Focusing this PR to the two fix only.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java [hunk 6]
```diff
@@ -1095,11 +1095,11 @@ public ListKeysResult listKeys(String volumeName, String bucketName,
     } else {
       // This allows us to seek directly to the first key with the right prefix.
       seekKey = getOzoneKey(volumeName, bucketName,
-          StringUtils.isNotBlank(keyPrefix) ? keyPrefix : OM_KEY_PREFIX);
+          StringUtils.isNotEmpty(keyPrefix) ? keyPrefix : OM_KEY_PREFIX);
```
Comment (Russole) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java:1098]: The `isNotBlank` -> `isNotEmpty` change looks correct for S3 prefix semantics, since whitespace/control-character prefixes are valid.

Should we add a dedicated OM/listKeys test for newline or whitespace prefixes? The removed endpoint test does not directly cover this code path.

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java (hunk 5 above)
Comment (Russole) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java]: That makes sense to me. Focusing this PR on the space encoding and empty delimiter fixes is reasonable.

The remaining `isNotBlank` -> `isNotEmpty` change also looks correct for S3 prefix semantics, since whitespace/control-character prefixes are valid. It may be worth adding a dedicated OM/listKeys test for that behavior, but I do not think it needs to block this PR.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java (hunk 6 above)
Comment (Gargi-jais11) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java:1098]: Thanks for pointing out this. Added the test case.

**Comment by Gargi-jais11** [comment]
@Russole Please take a look, I have updated the Pr.

**Review by Russole — APPROVED** [review]
Thanks @Gargi-jais11 for the patch. LGTM!

**Comment by Gargi-jais11** [comment]
Thanks @adoroszlai , @Russole and @chungen0126 for reviewing the patch.
